redaxo/src/addons/structure/boot.php

```diff
@@ -85,3 +88,518 @@
+                $type = match ($ep->getName()) {
+                    'ART_SLICES_COPY' => 'slices_copy',
+                    'SLICE_MOVE' => 'slice_' . $ep->getParam('direction'),
+                    default => strtolower($ep->getName()),
+                };
+
+                $articleId = $ep->getParam('article_id');
+                $clangId = $ep->getParam('clang_id');
+                $sliceRevision = $ep->getParam('slice_revision');
+
+                if (0 == $sliceRevision) {
+                    rex_article_slice_history::makeSnapshot($articleId, $clangId, $type);
+                }
+            },
+        );
+
+        rex_view::addCssFile($addon->getAssetsUrl('noUiSlider/nouislider.css'));
+        rex_view::addJsFile($addon->getAssetsUrl('noUiSlider/nouislider.js'), [rex_view::JS_IMMUTABLE => true]);
+        rex_view::addCssFile($addon->getAssetsUrl('history.css'));
+        rex_view::addJsFile($addon->getAssetsUrl('history.js'), [rex_view::JS_IMMUTABLE => true]);
+
+        switch (rex_request('rex_history_function', 'string')) {
+            case 'snap':
+                $articleId = rex_request('history_article_id', 'int');
+                $clangId = rex_request('history_clang_id', 'int');
+                $historyDate = rex_request('history_date', 'string');
+                rex_article_slice_history::restoreSnapshot($historyDate, $articleId, $clangId);
+
+                // no break
+            case 'layer':
+                $articleId = rex_request('history_article_id', 'int');
+                $clangId = rex_request('history_clang_id', 'int');
+                $versions = rex_article_slice_history::getSnapshots($articleId, $clangId);
+
+                $select1 = [];
+                $select1[] = '<option value="0" selected="selected" data-revision="0">' . $addon->i18n(
+                    'structure_history_current_version',
+                ) . '</option>';
+                if (true === $addon->getConfig('version', false)) {
+                    $select1[] = '<option value="1" data-revision="1">' . rex_i18n::msg(
+                        'version_workingversion',
+                    ) . '</option>';
+                }
+
+                $select2 = [];
+                $select2[] = '<option value="" selected="selected">' . $addon->i18n(
+                    'structure_history_current_version',
+                ) . '</option>';
+                foreach ($versions as $version) {
+                    $historyInfo = $version['history_date'];
+                    if ('' != $version['history_user']) {
+                        $historyInfo = $version['history_date'] . ' [' . $version['history_user'] . ']';
+                    }
+                    $select2[] = '<option value="' . strtotime(
+                        $version['history_date'],
+                    ) . '" data-history-date="' . rex_escape(
+                        $version['history_date'],
+                    ) . '">' . rex_escape($historyInfo) . '</option>';
```

Comment:
Irgendwie ist phpstorm bei dir suboptimal eingestellt. Also dass bei Copy/Paste der Code unschön neu formatiert wird. 
Sind hier einige Stellen in der Datei.

Reply:
Also du musst es jetzt hier nicht anpassen.
Evtl. korrigiere ich es noch bisschen.

Aber bevor du weitere Code-Blöcke kopierst, sollten wir evtl schauen, dass wir das bei dir besser einstellen in phpstorm.

Reply:
Habe die Umbrüche in boot.php angepasst